core/src/term/mod.rs

```diff
@@ -576,6 +607,34 @@ impl TypeAnnotation {
     pub fn is_empty(&self) -> bool {
         self.typ.is_none() && self.contracts.is_empty()
     }
+
+    /// **Warning**: the contract equality check used in this function behaves like syntactic
+    /// equality, and doesn't take the environment into account. It's unsound for execution (it
+    /// could equate contracts that are actually totally distinct), but we use it only to trim
+    /// accumulated contracts before pretty-printing. Do not use prior to any form of evaluation.
```

Comment:
Shouldn't this be sound because all the contracts are being applied in the same environment? There's no way for `foo | Foo | Foo` to sneak in a shadow of for `Foo` in the second contract application without being different syntactically.

If it's not sound, it seems problematic for pretty-printing as well. We've run into issues with the pretty-printer producing output that doesn't get parsed back the same way, right?